csrc/ir/utils.cpp

```diff
@@ -1310,4 +1310,102 @@ MmaOpDetails getMmaOpDetails(
+  const Val* in_a = mma_op->inA();
+  const Val* in_b = mma_op->inB();
+
+  NVF_CHECK(
+    in_a->as<TensorView>()->nDims() == in_b->as<TensorView>()->nDims(),
+    "Either both or none of A and B should be batch"
+  );
+  // Verify that the broadcasted size is 3.
+  NVF_CHECK(
+      in_a->as<TensorView>()->nDims() == 3,
+      "MmaOp::evaluate is not implemented for size: ",
+      in_a->as<TensorView>()->nDims());
+
+  NVF_CHECK(
+      in_a->definition() != nullptr &&
+          in_a->definition()->isA<BroadcastOp>(),
+      "Currently, MmaOp::evaluate assumes the preceding op to be a broadcast.");
+  NVF_CHECK(
+      in_b->definition() != nullptr &&
+          in_b->definition()->isA<BroadcastOp>(),
+      "Currently, MmaOp::evaluate assumes the preceding op to be a broadcast.");
+  NVF_CHECK(
+      in_a->as<TensorView>()->getRootDomain().back()->isBroadcast(),
+      "Expected last dimension to be broadcasted for first operand.");
+  NVF_CHECK(
+      in_b->as<TensorView>()->getRootDomain().front()->isBroadcast(),
+      "Expected first dimension to be broadcasted for second operand.");
+
+  // ATen preserves the dtype of MmaOp inputs whereas MmaOp generates float
+  // outputs. To preserve numerical equivalence and precision, the output of
+  // ATen matmul should be the same as MmaOp out `eventually`.
+  // See https://github.com/NVIDIA/Fuser/pull/1874#discussion_r1516991574
+  // Supported cases:
+  //  1. MmaOp->out() and MmaOp->input() are the same dtype.
+  //  2. MmaOp->out() is followed by a CastOp() to the MmaOp->input() dtype.
+  // NOTE: Currently MmaOp only accepts Half and BFloat16 so case (1) does not
+  // occur.
+
+  NVF_CHECK(in_a->as<TensorView>()->getDataType().value() == final_out_dtype,
+    "CastOp should cast to the same final datatype as the input datatype.");
+}
+
+void verifyBiasForEvaluation(
+  Val* bias,
+  DataType final_out_dtype) {
+
+  // Bias tensor is broadcasted from shape [M,] to [M, 1] in biasEpilogue
+  NVF_CHECK(bias->definition() != nullptr && bias->definition()->isA<BroadcastOp>(),
+        "Expected bias tensor to be broadcasted.");
+
+  NVF_CHECK(
+    bias->as<TensorView>()->getRootDomain().back()->isBroadcast(),
+    "Expected last dimension to be broadcasted for bias tensor.");
+
+  Val* bcast_in = bias->definition()->as<BroadcastOp>()->input(0);
+
+  // The bias tensor and matmul inputs should be of the same dtype.
+  // The bias tensor is upcasted to fp32.
+  auto is_previous_castop = [](Val* in, DataType expected_dtype) -> bool {
+    if (auto* unary = dynamic_cast<UnaryOp*>(in->definition())) {
+      if (unary->getUnaryOpType() == UnaryOpType::Cast) {
+        return unary->input(0)->getDataType().value() == expected_dtype;
+      }
+    }
+    return false;
+  };
+
+  NVF_CHECK(is_previous_castop(bcast_in, final_out_dtype),
+    "Expected the bias to be preceded by castOp before broadcasting and to be originally of the same type as the final output dtype .");
```

Comment:
Does addmm support different bias dtype? If so the can we make this a return value instead of an assertion, then handle this case gracefully?

Reply:
No. All the inputs have to be of the same type.